WordPress/gutenberg
# Components: Speak Notice message

Fixes #9442
Closes #15708 
Related: #15594

This pull request seeks to move spoken message handling for notices from the notices store into the `Notice` component.

**Implementation notes:**

_**Update (2019-12-10):** The commentary in this section has since received feedback at https://github.com/WordPress/gutenberg/pull/15745#issuecomment-495912383 , with subsequent revisions summarized at https://github.com/WordPress/gutenberg/pull/15745#issuecomment-564250328_

I could use some accessibility feedback here. Per #9442, there's a recommendation to use the `role` attribute for notices affected by `withNotices`. It's unclear whether this is appropriate for all notices. The current implementation here preserves existing behavior so far as using `wp.a11y.speak`, but notices (both existing and from `withNotices`) will default to a politeness level corresponding to the status of the message (error messages as `'assertive'`, success messages as `'polite'`). From the reading of the [relevant roles](https://www.w3.org/TR/wai-aria-1.1/#alert), it doesn't seem like this must be a guaranteed 1-to-1 mapping (i.e. a success notice could be a time-sensitive, assertive message?), but is implemented instead as a sensible default. Notices can be configured to override this default by assigning `role="polite"` or `role="assertive"`. The role is not in-fact assigned to the rendered element, but its usage with `speak` is assumed to be a suitable equivalent behavior.

Feedback needed:

- Is the default behavior sensible?
- Is the actual behavior for `role` in not assigning the DOM attribute acceptable?
   - In my testing, assigning it to the element would cause a duplicate spoken message, if `speak` is also used.
- Would it be better to always use `role` and never `wp.a11y.speak`?
- Should `role` be exposed at all, or instead based entirely from `status`? Or always `'assertive'`?

**Testing instructions:**

Verify there are no regressions in the spoken messages of notices (example: published post "Post published!" notice).

Verify that notices mounted by `withNotices` (or ad hoc use of the `Notice` component) are spoken.
